src/engines/nwn2/inventory.cpp

```diff
@@ -101,6 +121,28 @@ void Inventory::clear() {
 	inventoryItems.clear();
 }
 
+uint16 Inventory::getFirstEmptySlot() const {
+	// Empty inventory?
+	const size_t total = inventoryItems.size();
+	if (total == 0)
+		return 0;
+
+	// Check for no open slots
+	const uint16 max = inventoryItems.rbegin()->first;
+	if (max < total)
+		return total;
+
+	// Look for the first open slot
+	uint16 slot;
+	std::map<uint16, Item *>::const_iterator item;
+	for (item = inventoryItems.begin(), slot = 0; item != inventoryItems.end(); ++item, slot++)
+		if (slot < item->first)
+			return slot;
```

Comment:
What exactly is that doing? That looks...confused.

Should the inventory be a `std::map<uint16, Item *>` in the first place? Wouldn't it make more sense to have a flat `std::vector<Item *>`, of the size of the inventory, with the item being nullptr if unused?

(Or maybe even a `std::vector<std::unique_ptr<Item>>`, or `std::vector<std::shared_ptr<Item>>`, if the inventory is supposed to manage the item.)

Finding the first non-used inventory slot would then be a `std::find_if`. You could even just return the iterator on this method (and that would also give you a way to express "no free slot", by returning the end iterator). If you need the index later (because a script call wants it or something), you can use `std::distance()` then.

Reply:
That's my attempt to optimize the search for a gap in the map.

1) Most creatures are going to have an empty inventory, so it checks that right away.
2) The next case is for inventory with no gaps, which will be true for most creatures as well as for PC's that have sorted their inventory. In that case just checking the length is enough to see if there are gaps.
3) Finally, I do the slow crawl through the list looking for the first gap.

The PC inventory could potentially have a lot of gaps, so I just choose map for optimum storage. But yes it could use a vector, if you don't mind the large amount of wasted storage. I suppose I could allocate a row at a time.

Reply:
I'd say optimizing for a large amount of gaps here would be premature optimization. Especially considering the usual number of items in an inventory and that you're giving up cache locality with a map. I'm not really sure such a map would even be faster than a vector here, to be frank.

When we see that crawling through the inventory is a bottle neck, we can optimize it further. Preferably with some profiling, so that we can see how different structures perform.

Reply:
Okay, I will simplify it down. Thanks.